### PR TITLE
Require prefix conditions to start with a /

### LINKED ISSUE #1628: All prefix: conditions must start with a /
Breaking out from https://github.com/projectcontour/contour/issues/1597#issuecomment-537754073

I propose that all `prefix: ` conditions, if present, must start with a `/`. Specifically

- At most one `prefix: ` condition may be present in each `include.conditions` or `route.conditions` block.
- If present the `prefix: ` condition must start with a `/`.

### PR DESCRIPTION
**Pull request by youngnick**
Fixes #1628
Updates #1597

Signed-off-by: Nick Young <ynick@vmware.com>

### CODE REVIEW

File: internal/dag/conditions.go [hunk 1]
```diff
@@ -30,13 +31,21 @@ func pathCondition(conds []projcontour.Condition) Condition {
 	}
 }
 
-func pathConditionsValid(conds []projcontour.Condition) bool {
+// pathConditionsValid validates a slice of Conditions.
+// It encodes the business rules about what is allowed for prefix Conditions.
+func pathConditionsValid(conds []projcontour.Condition, sw *ObjectStatusWriter, conditionsContext string) bool {
```
Comment (youngnick) [inline internal/dag/conditions.go]: I'm not a big fan of passing this conditionsContext string, but without it there is no way at all to tell which conditions block is invalid in a HTTPProxy document. Solving this properly requires solving #1575 as well, so this is here for now.

File: internal/dag/builder.go [hunk 2]
```diff
@@ -517,8 +517,7 @@ func (b *Builder) computeRoutes(sw *ObjectStatusWriter, proxy *projcontour.HTTPP
 				return nil
 			}
 
-			if !pathConditionsValid(include.Conditions) {
-				sw.SetInvalid("include: cannot specify multiple path conditions in the same include")
+			if !pathConditionsValid(include.Conditions, sw, "include") {
```
Comment (davecheney) [inline internal/dag/builder.go]: suggestion: for consistency, please move sw to the front of the parameter list.

File: internal/dag/conditions.go [hunk 3]
```diff
@@ -30,13 +31,21 @@ func pathCondition(conds []projcontour.Condition) Condition {
 	}
 }
 
-func pathConditionsValid(conds []projcontour.Condition) bool {
+// pathConditionsValid validates a slice of Conditions.
```
Comment (davecheney) [inline internal/dag/conditions.go]: ... validates a slice of conditions can be correctly merged.

File: internal/dag/conditions.go [hunk 4]
```diff
@@ -30,13 +31,21 @@ func pathCondition(conds []projcontour.Condition) Condition {
...
+// It encodes the business rules about what is allowed for prefix Conditions.
+func pathConditionsValid(conds []projcontour.Condition, sw *ObjectStatusWriter, conditionsContext string) bool {
 	found := 0
 	for _, cond := range conds {
 		if cond.Prefix != "" {
+			// We found a prefix, so increment the prefix counter.
```
Comment (davecheney) [inline internal/dag/conditions.go]: I don't think this commend adds a lot. Perhaps something like renaming `found` to `prefixCount` would make the code more self explanatory.

File: internal/dag/conditions_test.go [hunk 5]
```diff
@@ -270,11 +271,37 @@ func TestConditionsValid(t *testing.T) {
...
 	}
 
 	for name, tc := range tests {
 		t.Run(name, func(t *testing.T) {
-			got := pathConditionsValid(tc.conditions)
+			// TODO(youngnick) This feels dirty but I don't know a better solution.
+			// We're not testing the returned info from the ObjectStatusWriter here,
+			// that's done at the status_test.go layer.
+			swblank := &ObjectStatusWriter{}
+			sw, _ := swblank.WithObject(&projcontour.HTTPProxy{
```
Comment (davecheney) [inline internal/dag/conditions_test.go]: yeah this is a bit unfortunate. Perhaps in the future status writer should become an interface then we can supply a dummy.

File: internal/dag/conditions.go (hunk 4 above)
Comment (youngnick) [reply internal/dag/conditions.go]: I actually started doing that when I was rewriting into a single verify function, and forgot to bring it back when I reverted. Agreed, I'll do that now.

File: internal/dag/conditions_test.go (hunk 5 above)
Comment (youngnick) [reply internal/dag/conditions_test.go]: Ah, that makes sense. Consider my mental note made, but I'll put a TODO there and make an issue.

File: internal/dag/builder.go (hunk 2 above)
Comment (youngnick) [reply internal/dag/builder.go]: Missed that, no problem.

File: internal/dag/conditions.go [hunk 6]
```diff
@@ -30,13 +31,20 @@ func pathCondition(conds []projcontour.Condition) Condition {
...
-	found := 0
+// pathConditionsValid validates a slice of Conditions can be correctly merged.
+// It encodes the business rules about what is allowed for prefix Conditions.
+func pathConditionsValid(sw *ObjectStatusWriter, conds []projcontour.Condition, conditionsContext string) bool {
+	prefixCount := 0
 	for _, cond := range conds {
 		if cond.Prefix != "" {
-			found++
+			prefixCount++
+			if !strings.HasPrefix(cond.Prefix, "/") {
```
Comment (davecheney) [inline internal/dag/conditions.go]: suggestion: we know that cond.Prefix != "" at this point so you can just write
```
if cond.Prefix[0] != '/' {
```

Comment (davecheney) [reply internal/dag/conditions.go]: more correctly len(cond.Prefix) > 0 so its safe to access subscript 0

Comment (youngnick) [reply internal/dag/conditions.go]: of course. Neat.

**Pull request merged**